packages/host/app/components/operator-mode/code-editor.gts

```diff
@@ -0,0 +1,310 @@
+    onSetup: (
+      updateCursorByDeclaration: (declaration: ModuleDeclaration) => void,
+    ) => void;
+  };
+}
+
+const log = logger('component:code-editor');
+
+export default class CodeEditor extends Component<Signature> {
+  @service declare monacoService: MonacoService;
+  @service declare operatorModeStateService: OperatorModeStateService;
+  @service declare cardService: CardService;
+
+  @tracked private maybeMonacoSDK: MonacoSDK | undefined;
+
+  private hasUnsavedSourceChanges = false;
+
+  constructor(owner: Owner, args: Signature['Args']) {
+    super(owner, args);
+
+    registerDestructor(this, () => {
+      // destructor functons are called synchronously. in order to save,
+      // which is async, we leverage an EC task that is running in a
+      // parent component (EC task lifetimes are bound to their context)
+      // that is not being destroyed.
+      if (this.codePath && this.hasUnsavedSourceChanges) {
+        let monacoContent = this.monacoService.getMonacoContent();
+        if (monacoContent) {
+          this.args.saveSourceOnClose(this.codePath, monacoContent);
+        }
+      }
+    });
+
+    this.loadMonaco.perform();
+
+    this.args.onSetup(this.updateMonacoCursorPositionByDeclaration);
+  }
+
+  private get isReady() {
+    return this.maybeMonacoSDK && isReady(this.args.file);
+  }
+
+  private get isLoading() {
+    return (
+      this.loadMonaco.isRunning || this.args.moduleContentsResource?.isLoading
+    );
+  }
+
+  private get declarations() {
+    return this.args.moduleContentsResource?.declarations || [];
+  }
+
+  private loadMonaco = task(async () => {
+    this.maybeMonacoSDK = await this.monacoService.getMonacoContext();
+  });
+
+  private get monacoSDK() {
+    if (this.maybeMonacoSDK) {
+      return this.maybeMonacoSDK;
+    }
+    throw new Error(`cannot use monaco SDK before it has loaded`);
+  }
+
+  private get codePath() {
+    return this.operatorModeStateService.state.codePath;
+  }
+
+  private get readyFile() {
+    if (isReady(this.args.file)) {
+      return this.args.file;
+    }
+    throw new Error(
+      `cannot access file contents ${this.codePath} before file is open`,
+    );
+  }
+
+  @action
+  private initializeMonacoCursorPosition() {
+    if (this.args.selectedDeclaration?.path?.node.loc) {
+      let { start } = this.args.selectedDeclaration.path.node.loc;
+      this.monacoService.updateCursorPosition(
+        new Position(start.line, start.column),
+      );
+    }
+  }
+
+  @action
+  private updateMonacoCursorPositionByDeclaration(
+    declaration: ModuleDeclaration,
+  ) {
+    if (declaration.path?.node.loc) {
+      let { start, end } = declaration.path?.node.loc;
+      let currentCursorPosition = this.monacoService.getCursorPosition();
+      if (
+        currentCursorPosition &&
+        (currentCursorPosition.lineNumber < start.line ||
+          currentCursorPosition.lineNumber > end.line)
+      ) {
```

Comment:
Does this mean that if there is no cursor position, then the declaration does get the cursor position moved to it? Or is there always a cursor position?